src/pudl/scripts/dbt_helper.py

```diff
@@ -201,6 +201,76 @@ def to_yaml(self, schema_path: Path):
             yaml_output = _prettier_yaml_dumps(self.model_dump(exclude_none=True))
             schema_file.write(yaml_output)
 
+    def merge_metadata_from(self, old_schema: "DbtSchema") -> "DbtSchema":
+        """Merge metadata from an old schema into this one, preferring new values where present."""
+
+        def merge_tables(new_tables, old_tables):
+            old_table_map = {t.name: t for t in old_tables or []}
+            merged = []
+
+            for new_table in new_tables or []:
+                old_table = old_table_map.get(new_table.name)
+                if not old_table:
+                    merged.append(new_table)
+                    continue
+
+                old_columns = {c.name: c for c in old_table.columns or []}
+                merged_columns = [
+                    (
+                        col.model_copy(
+                            update={
+                                "description": col.description
+                                or old_columns.get(col.name, {}).description,
+                                "data_tests": col.data_tests
+                                or old_columns.get(col.name, {}).data_tests,
+                                "tags": col.tags or old_columns.get(col.name, {}).tags,
+                            }
+                        )
+                        if col.name in old_columns
+                        else col
+                    )
+                    for col in new_table.columns or []
+                ]
+
+                merged_table = new_table.model_copy(
+                    update={
+                        "description": new_table.description or old_table.description,
+                        "data_tests": new_table.data_tests or old_table.data_tests,
+                        "tags": new_table.tags or old_table.tags,
+                        "meta": new_table.meta or old_table.meta,
+                        "columns": merged_columns,
+                    }
+                )
+                merged.append(merged_table)
+
+            return merged
+
+        # Merge sources
+        old_sources = {s.name: s for s in old_schema.sources or []}
+        new_sources = []
+        for new_source in self.sources or []:
+            old_source = old_sources.get(new_source.name)
+            merged_tables = merge_tables(
+                new_source.tables, old_source.tables if old_source else []
+            )
+            merged_source = new_source.model_copy(update={"tables": merged_tables})
+            new_sources.append(merged_source)
+
+        # Merge models (if any)
+        if self.models:
```

Comment:
I suppose we can't *define* any models in `pudl.metadata.resources` so maybe we only need `self.model_copy(update={"sources": new_sources})` and can get rid of this whole block.

Reply:
I guess you could arrange to try and merge an old model with another old model -- reading both of them out of `schema.yml` files, rather than generating one from the PUDL metadata. But I don't think that's a scenario we're trying to support. So maybe we can move the `if self.models` up to the top and have it return an error early.

Reply:
Fair - we *currently* don't ever call `merge_metadata_from` on any schemas that we read from YAML, but that doesn't mean we can't in the future.

I would personally make `merge_metadata_from` only take the sources as a parameter, instead of the whole schema, so that we can't accidentally pass in models at all. But checking at the top achieves a similar effect with less refactoring so it's probably a smarter move.

Reply:
I have removed support for DBT models in the schema merging logic and added a ValueError in case someone tries to merge a schema with models defined.